blocks/algorithms/__init__.py

```diff
@@ -222,6 +224,29 @@ def compute_step(self, param, grad_wr_param):
         """
         raise NotImplementedError
 
+    def compute_steps(self, gradients):
+        """Build a Theano expression for steps for all parameters.
+
+        Override this method if you want to process the gradient
+        with respect to all parameters as a whole, not parameter-wise.
+
+        Parameters
+        ----------
+        gradients : dict of (:class:`~tensor.TensorSharedVariable`
```

Comment:
I will surely do (a), but as for (b), I think it is an unnecessary complication of the step rule concept. Step rules live at a high level of abstraction where the order of computations does not matter and simplify provide the user with the expressions he needs, and it is up to him how exactly he wants arrange these computations.